### PR TITLE
Simplify NativeIntegerAttribute encoding

### LINKED ISSUE #44086: Consider streamlining flag encoding for native integers
Looking at `NativeIntegerTransformsEncoder`, native int encoding doesn't encode a flag for modreq types, but it does encode a flag for types that are not `IntPtr`. 
This seems a strange middle-ground between the tuple approach (only encode a flag for types that are relevant, ie. ValueTuple) and the dynamic approach (encode a flag for all types, including modreqs).

Tagging @cston @gafter 

Relates to https://github.com/dotnet/roslyn/issues/38821 (test plan for native ints)

### PR DESCRIPTION
**Pull request by cston**
Encode a `bool` value for each `System.IntPtr` and `System.UIntPtr` only, and ignore custom modifiers.

Fixes #44086.

### CODE REVIEW

**Comment by jcouv** [comment]
Is there a corresponding update to design/spec doc?

**Comment by jaredpar** [comment]
Do we have a write up for what the encoding is here so that others can use it? Similar to what we have for NRT

https://github.com/dotnet/roslyn/blob/master/docs/features/nullable-metadata.md

**Comment by cston** [comment]
> Do we have a write up for what the encoding is here 

Added docs/features/NativeIntegerAttribute.md to the PR.

File: docs/features/NativeIntegerAttribute.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+The `bool[]` is constructed as follows:
+- `System.IntPtr`, `System.UIntPtr`: `true` if the type represents a native integer; `false` otherwise
+- Generic type: the representation of the type arguments in order including containing types
+- Tuple: the representation of the underlying constructed type
+- Array: the representation of the element type
+- Pointer type: the representation of the pointed at type
+- Function pointer type: the representation of the return type followed by the parameter types in order
+- Other: none
+
+Given the construction above, the number of elements in the `bool[]` matches the number of references to `System.IntPtr` and `System.UIntPtr` in the type.
```
Comment (333fred) [inline docs/features/NativeIntegerAttribute.md:54]: Consider clarifying whether modopts/reqs are included.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_NativeInteger.cs [hunk 2]
```diff
@@ -210,35 +210,344 @@ public void MissingAttributeUsageAttribute()
...
+.class public A<T, U>
+{
+  .method public hidebysig specialname rtspecialname instance void .ctor() cil managed { ret }
+}
+.class public B
+{
+  .method public static void F0(native int x, native uint y)
+  {
+    .param [1]
+    .custom instance void System.Runtime.CompilerServices.NativeIntegerAttribute::.ctor(bool[]) = ( 01 00 00 00 00 00 00 00 )
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_NativeInteger.cs:231]: Consider adding a comment to clarify what these arrays actually contain.

**Comment by 333fred** [comment]
Done review pass (commit 2). I didn't fully review the tests, as I'm unsure what the IL actually represents.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_NativeInteger.cs [hunk 3]
```diff
@@ -210,35 +210,344 @@ public void MissingAttributeUsageAttribute()
...
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "nuint").WithArguments("native-sized integers").WithLocation(6, 25),
+                // (7,20): error CS8652: The feature 'native-sized integers' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         B.F2(new A<nint, uint>());
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "nint").WithArguments("native-sized integers").WithLocation(7, 20));
+
+            comp = CreateCompilation(source1, new[] { ref0 }, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+
+            var type = comp.GetTypeByMetadataName("B");
+            Assert.Equal("void B.F0(nint x, nuint y)", type.GetMember("F0").ToDisplayString(FormatWithSpecialTypes));
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_NativeInteger.cs:420]: Consider verifying these on both preview and 8 compilations.

**Comment by RikkiGibson** [comment]
@cston is this ready to merge?

**Comment by cston** [comment]
> Is there a corresponding update to design/spec doc?

[https://github.com/dotnet/csharplang/pull/3856](https://github.com/dotnet/csharplang/pull/3856)

---
In reply to: [668256851](https://github.com/dotnet/roslyn/pull/46522#issuecomment-668256851) [](ancestors = 668256851)

**Pull request merged**